core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,331 @@
+
+final class RefreshingAddressResolver extends AbstractAddressResolver<InetSocketAddress> {
+
+    private static final Logger logger = LoggerFactory.getLogger(RefreshingAddressResolver.class);
+
+    private final ConcurrentMap<String, CompletableFuture<CacheEntry>> cache;
+    private final DefaultDnsNameResolver resolver;
+    private final List<DnsRecordType> dnsRecordTypes;
+    private final int minTtl;
+    private final int maxTtl;
+    private final int negativeTtl;
+    private final Backoff refreshBackoff;
+
+    private volatile boolean resolverClosed;
+
+    RefreshingAddressResolver(EventLoop eventLoop,
+                              ConcurrentMap<String, CompletableFuture<CacheEntry>> cache,
+                              DefaultDnsNameResolver resolver, List<DnsRecordType> dnsRecordTypes,
+                              int minTtl, int maxTtl, int negativeTtl, Backoff refreshBackoff) {
+        super(eventLoop);
+        this.cache = cache;
+        this.resolver = resolver;
+        this.dnsRecordTypes = dnsRecordTypes;
+        this.minTtl = minTtl;
+        this.maxTtl = maxTtl;
+        this.negativeTtl = negativeTtl;
+        this.refreshBackoff = refreshBackoff;
+    }
+
+    @Override
+    protected boolean doIsResolved(InetSocketAddress address) {
+        return !requireNonNull(address, "address").isUnresolved();
+    }
+
+    @Override
+    protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise)
+            throws Exception {
+        requireNonNull(unresolvedAddress, "unresolvedAddress");
+        requireNonNull(promise, "promise");
+        if (resolverClosed) {
+            promise.tryFailure(new IllegalStateException("resolver is closed already."));
+            return;
+        }
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
+        final CompletableFuture<CacheEntry> entryFuture = cache.get(hostname);
+        if (entryFuture != null) {
+            handleFromCache(entryFuture, promise, port);
+            return;
+        }
+
+        final CompletableFuture<CacheEntry> result = new CompletableFuture<>();
+        final CompletableFuture<CacheEntry> previous = cache.putIfAbsent(hostname, result);
+        if (previous != null) {
+            handleFromCache(previous, promise, port);
+            return;
+        }
+
+        final List<DnsQuestion> questions =
+                dnsRecordTypes.stream()
+                              .map(type -> DnsQuestionWithoutTrailingDot.of(hostname, type))
+                              .collect(toImmutableList());
+        sendQueries(questions, hostname, result);
+        result.handle((entry, unused) -> {
+            final Throwable cause = entry.cause();
+            if (cause != null) {
+                if (negativeTtl > 0) {
+                    executor().schedule(() -> cache.remove(hostname), negativeTtl, TimeUnit.SECONDS);
```

Comment:
I think Negative TTL is a kind of cache. `RefreshingAddressResolver` role is keeping cache and automatic refresh it. So the following logic will work to meet the feature that refreshes the failed result with `nagativeTtl`. 
```java
entry.scheduleRefresh(TimeUnit.SECONDS.toMillis(negativeTtl);
```

If the next result is successful, it will eventually call  `entry.scheduleRefresh(entry.ttlMillis())` next time. WDYT?

Reply:
https://github.com/line/armeria/pull/2217#issuecomment-547723266
Please see this.